Detector/DetCommon/src/DetUtils.cpp

```diff
@@ -11,10 +23,160 @@ DD4hep::XML::Component det::getNodeByStrAttr(const DD4hep::XML::Handle_t& mother
   return DD4hep::XML::Component(nullptr);
 }
 
-double det::getAttrValueWithFallback(const DD4hep::XML::Component& node, const std::string& attrName, const double& defaultValue) {
+double getAttrValueWithFallback(const DD4hep::XML::Component& node, const std::string& attrName, const double& defaultValue) {
   if (node.hasAttr(attrName.c_str())) {
     return node.attr<double>(attrName.c_str());
   } else {
     return defaultValue;
   }
 }
+
+uint64_t cellID(const DD4hep::Geometry::Segmentation& aSeg, const G4Step& aStep, bool aPreStepPoint) {
+  DD4hep::Simulation::Geant4VolumeManager volMgr =
+    DD4hep::Simulation::Geant4Mapping::instance().volumeManager();
+  DD4hep::Geometry::VolumeManager::VolumeID volID =
+    volMgr.volumeID(aStep.GetPreStepPoint()->GetTouchable());
+  if (aSeg.isValid() )  {
+    G4ThreeVector global;
+    if (aPreStepPoint) {
+      global = aStep.GetPreStepPoint()->GetPosition();
+    } else {
+      global = 0.5*(aStep.GetPreStepPoint()->GetPosition()+aStep.GetPostStepPoint()->GetPosition());
+    }
+    G4ThreeVector local  = aStep.GetPreStepPoint()->GetTouchable()->
+      GetHistory()->GetTopTransform().TransformPoint(global);
+    DD4hep::Geometry::Position loc(local.x()*MM_2_CM, local.y()*MM_2_CM, local.z()*MM_2_CM);
+    DD4hep::Geometry::Position glob(global.x()*MM_2_CM, global.y()*MM_2_CM, global.z()*MM_2_CM);
+    DD4hep::Geometry::VolumeManager::VolumeID cID = aSeg.cellID(loc,glob,volID);
+    return cID;
+  }
+  return volID;
+}
+
+std::vector<uint64_t> neighbours(DD4hep::DDSegmentation::BitField64& aDecoder,
+    const std::vector<std::string>& aFieldNames,
+    const std::vector<std::pair<int,int>>& aFieldExtremes,
+    uint64_t aCellId) {
+  std::vector<uint64_t> neighbours;
+  aDecoder.setValue(aCellId);
+  for(uint itField=0; itField < aFieldNames.size(); itField++) {
+    const auto& field = aFieldNames[itField];
+    int id = aDecoder[field];
+    if(id > aFieldExtremes[itField].first) {
+      aDecoder[field] = id-1;
+      neighbours.emplace_back(aDecoder.getValue());
+    }
+    if(id < aFieldExtremes[itField].second) {
+      aDecoder[field] = id+1;
+      neighbours.emplace_back(aDecoder.getValue());
+    }
+    aDecoder[field] = id;
+  }
+  return std::move(neighbours);
+}
+
+std::vector<std::pair<int,int>> bitfieldExtremes(DD4hep::DDSegmentation::BitField64& aDecoder,
+    const std::vector<std::string>& aFieldNames) {
+  std::vector<std::pair<int,int>> extremes;
+  int width = 0;
+  for(const auto& field: aFieldNames) {
+    width = aDecoder[field].width();
+    if( aDecoder[field].isSigned() ){
+      extremes.emplace_back(std::make_pair( ( 1LL << ( width - 1 ) ) - ( 1LL << width ),
+          ( 1LL << ( width - 1 ) ) - 1));
+    } else {
+      extremes.emplace_back(std::make_pair( 0,
+         (0x0001 << width) - 1 ));
```

Comment:
Although, thinking about this a bit more:
1) `1LL` is better because it explicitly notes the size
2) But do we actually expect 64 bit here? Since this is the individual fields, I assume it's possible, but unlikely?
3) If we work with `long long` the return type should reflect that